File: rust/checker.rs
use rayon::prelude::*;

pub fn start_checking(filename: &str, min_set_size: u8, max_set_size: u8) {
    for set_size in min_set_size..=max_set_size {
        let total_relations = 1 << (set_size * set_size);
        let start = std::time::Instant::now();
        let count = check_relations(set_size, total_relations);
        let elapsed = start.elapsed().as_secs_f64();

        crate::summary::print_results(set_size, count, total_relations, elapsed);
        crate::summary::append_results_to_csv(filename, set_size, count, total_relations, elapsed);
    }
}

pub fn check_relations(
    set_size: u8,
    total_relations: u64,
) -> [u64; crate::constants::NUMBER_OF_PROPERTIES] {
    let counts = (0..total_relations)
        .into_par_iter()
        .map(|i| {
            let mut relation = [0u8; crate::constants::MAX_SET_SIZE];
            let mut val = i;
            for j in 0..set_size {
                relation[j as usize] = (val & ((1 << set_size) - 1)) as u8;
                val >>= set_size;
            }
            let mut local = [0u64; crate::constants::NUMBER_OF_PROPERTIES];
            for j in 0..crate::constants::NUMBER_OF_PROPERTIES {
                local[j] = crate::properties::PROPERTY_FUNCTIONS[j](relation, set_size) as u64;
            }
            local
        })
        .reduce(
            || [0u64; crate::constants::NUMBER_OF_PROPERTIES],
            |mut a, b| {
                for i in 0..crate::constants::NUMBER_OF_PROPERTIES {
                    a[i] += b[i];
                }
                a
            },
        );
    counts
}

File: highly_optimized/src/checker.c
#include "checker.h"

#include <inttypes.h>
#include <omp.h>
#include <stdint.h>
#include <stdio.h>
#include <time.h>

#include "constants.h"
#include "properties.h"
#include "summary.h"

void start_checking(const char filename[], uint8_t min_set_size,
                    uint8_t max_set_size) {
    for (uint8_t set_size = min_set_size; set_size <= max_set_size;
         set_size++) {
        printf("Checking relations on set size: %" PRIu8 "\n", set_size);

        uint64_t total_relations = 1ULL << (set_size * set_size);
        uint64_t count[NUMBER_OF_PROPERTIES] = {0};

        clock_t start = clock();

        check_relations(set_size, total_relations, count);

        clock_t end = clock();
        float elapsed = (float)(end - start) / CLOCKS_PER_SEC;

        append_results_to_csv(filename, set_size, count, total_relations,
                              elapsed);
    }
}

void check_relations(uint8_t set_size, uint64_t total_relations,
                     uint64_t count[NUMBER_OF_PROPERTIES]) {
#pragma omp parallel
    {
        uint64_t thread_counts[NUMBER_OF_PROPERTIES] = {0};
        uint8_t relation[MAX_SET_SIZE];

#pragma omp for schedule(static)
        for (uint64_t idx = 0; idx < total_relations; idx++) {
            uint64_t val = idx;

            for (uint8_t i = 0; i < set_size; i++) {
                relation[i] = val & ((1 << set_size) - 1);
                val >>= set_size;
            }

            for (uint8_t i = 0; i < NUMBER_OF_PROPERTIES; i++)
                thread_counts[i] += PROPERTY_FUNCTIONS[i](relation, set_size);
        }

#pragma omp critical
        for (uint8_t i = 0; i < NUMBER_OF_PROPERTIES; i++)
            count[i] += thread_counts[i];
    }
}
